packages/graphql/src/schema/validation/validate-document.ts

```diff
@@ -180,15 +181,24 @@ function validateDocument(
         ],
     });
 
-    const schema = extendSchema(schemaToExtend, doc);
-
-    const errors = validateSchema(schema);
-
-    const filteredErrors = errors.filter((e) => e.message !== "Query root type must be provided.");
+    return extendSchema(schemaToExtend, doc, { assumeValid: !validateTypeDefs });
+}
 
-    if (filteredErrors.length) {
-        throw new Error(filteredErrors.join("\n"));
+function validateDocument(
+    document: DocumentNode,
+    validateTypeDefs = true,
+    additionalDirectives: Array<GraphQLDirective> = [],
+    additionalTypes: Array<GraphQLNamedType> = []
+): GraphQLSchema {
+    const schema = getBaseSchema(document, validateTypeDefs, additionalDirectives, additionalTypes);
+    if (validateTypeDefs) {
+        const errors = validateSchema(schema);
+        const filteredErrors = errors.filter((e) => e.message !== "Query root type must be provided.");
+        if (filteredErrors.length) {
+            throw new Error(filteredErrors.join("\n"));
+        }
     }
+    return schema;
```

Comment:
Sorry to go around in circles on this one, but I think best for now to keep all validation within this `validateDocument` function to save the quite hacky (IMO) return of the schema. You already have the boolean logic in here, so this will work well!

```suggestion
    validateCustomResolverRequires(document, schema);
```